<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [1]:
import pandas as pd
import psycopg2
import warnings
warnings.filterwarnings('ignore')

In [2]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python 
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [3]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [4]:
# текст запроса
query_3_1 = f'''
select count(*) as cnt_vacancies
from vacancies
'''

In [5]:
# результат запроса
df1 = pd.read_sql_query(query_3_1,connection)
print('В базе представлено, ',df1.iloc[0][0],' вакансий.')

В базе представлено,  49197  вакансий.


2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [6]:
# текст запроса
query_3_2 = f'''
select count(*) as cnt_employers
from employers
'''

In [7]:
# результат запроса
df2 = pd.read_sql_query(query_3_2,connection)
print('База данных "employers" содержит информацию о ',df2.iloc[0][0], ' работодателях.')

База данных "employers" содержит информацию о  23501  работодателях.


3. Посчитайте с помощью запроса количество регионов (таблица areas).

In [8]:
# текст запроса
query_3_3 = f'''
select count(*) as cnt_areas
from areas
'''

In [9]:
# результат запроса
df3 = pd.read_sql_query(query_3_3,connection)
print ('В базе данных "areas" представлено',df3.iloc[0][0], 'региона.') 

В базе данных "areas" престалено 1362 региона.


4. Посчитайте с помощью запроса количество сфер деятельности в базе (таблица industries).

In [10]:
# текст запроса
query_3_4 = f'''
select count(id) as cnt_industries
from industries
'''

In [11]:
# результат запроса
df4 = pd.read_sql_query(query_3_4,connection)
df4 
print ('В базе данных "industries" содержится',df4.iloc[0][0], 'сферы деятельности.') 

В базе данных "industries" содержится 294 сферы деятельности.


***

<span style="color:#59afe1"> 

# Выводы по предварительному анализу данных:


Само по себе числа вакансий, компаний, регионов и сфер деятельности не о чем не говорят.  
Для анализа этих показатель их нужно с чем-то сравнить.

В зависимости от задачи исследования, можно выявить только те вакансии, которые относятся к России   
или наоборот, проанализировать рынок зарубежных компаний.

HeadHunter крупная база вакансий в России, но есть и другие базы. Изучив все эти базы можно было бы   
понять на сколько обширна представлена анализируемая база. Что касается зарубежный рынков, то я уверен,   
что есть базы куда крупнее чем Head Hunter, Например, Linkedin. Данные из это базы можно было бы  
сравнить с частью зарубежных компаний БД HeadHunter и понять какие компании размещают свою резюме  
и в зарубежных и в Российский базах данных.

Число регионов и сфер деятельности было бы интересно сравнить с общим числом этих показателей в  
странах представленных в анализируемой БД. Таким образом можно понять на сколько широкую географию  
охватывает БД и сколько возможностей дает соискателю. Например если вакансии в БД представлены в основном  
западными регионам, то соискатели из Дальнего Востока могут попробовать искать вакансии в другом месте.  
Или, проанализировав спрос на вакансии в ДВ регионе, можно предложить компании Head Hunter увеличить число  
вакансий на востоке страны.

Можно выяснить сколько вакансий приходится в среднем на одну зарегистрированную компанию в стране.  
И в зависимости от более глубокой аналитики выяснить является это кадровым "голодом" или нет.
   
Мой общий вывод по Ютину 3 таков: 
Само по себе числа вакансий компаний, регионов и сфер деятельности не о чем не говорят.  
Для анализа этих показатель их нужно с чем-то сравнить.

# Юнит 4. Детальный анализ вакансий

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [12]:
# текст запроса
query_4_1 = f'''
-- для получения результата запроса необходимо объединить две таблицы БД: "areas" и "vacancies".
-- чтобы соотнести количество вакансий с конкретным регионом необходимо произвести объединение таблиц по ключу area_id

-- в качестве ответа на запрос получим названия регионов из столбца "name" таблицы "areas"
-- добавим к ним информацию о количестве вакансий в этом регионе, указанных в таблице "vacancies"
select 
    a.name as Город,
    count(v.id) as Число_вакансий
from vacancies as v
join areas as a on v.area_id = a.id

-- для корректной работы запроса необходимо сгруппировать данные по признаку "area_id"
group by 
    a.id
order by 2 DESC
'''

In [13]:
# результат запроса
df_4_1 = pd.read_sql_query(query_4_1,connection)
print('Топ 10 регионов России по количеству вакансий')
df_4_1.head(10)

Топ 10 регионов России по количеству вакансий


Город  Число_вакансий
0           Москва            5333
1  Санкт-Петербург            2851
2            Минск            2112
3      Новосибирск            2006
4           Алматы            1892
5     Екатеринбург            1698
6  Нижний Новгород            1670
7           Казань            1415
8        Краснодар            1301
9           Самара            1144

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [14]:
# текст запроса
query_4_2 = f'''
-- для получения необходимой информации воспользуемся таблице БД "vacancies"
-- выберем из таблицы количество вакансий для которых указана хотя бы одна из зарплатных вилок
-- для того чтобы ответ на SQL-запрос был более информативным, добавим общее количество вакансий и выведем результат в виде таблицы

-- в первой строке таблицы получим информацию о количестве в которых заполнено хотя бы одно из двух полей с зарплатой 
select 
    'cnt_not_null_salary' as type,
    count(*) as cnt
from vacancies as v

-- пропишем условие для отбора не пустых ячеек
where
    v.salary_from is not null or v.salary_to is not null

-- для объединения двух строк используем операцию UNION
union

-- вторая строка таблицы содержит информацию о общем количестве вакансий представленных в БД
select 
    'total_cnt',
    count(*) as cnt
from vacancies as v
join areas as a on v.area_id = a.id
'''

In [15]:
# результат запроса
df_4_2 = pd.read_sql_query(query_4_2,connection)
print('В базе данных "vacancies"',df_4_2.iloc[0][1],'вакансии имеют информацию о верхнем или нижнем уровне зарплаты.')
print('Это составляет',round(100*df_4_2.iloc[0][1]/df_4_2.iloc[1][1]),'% от общего количества вакансий представленных в БД')


В базе данных "vacancies" 24073 вакансии имеют информацию о верхнем или нижнем уровне зарплаты.
Это составляет 49 % от общего количества вакансий представленных в БД


3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [16]:
# текст запроса

query_4_3 = f'''
-- для получения необходимой информации воспользуемся таблице БД "vacancies"
-- посчитаем среднее значение для каждой зарплатной вилки с помощью агрегирующей функции avg
-- для информативности ответа на sql-запрос добавим данные о наименьшем и наибольшем значение для зарплатной вилки
-- учтем, что пустые строки (Null) игнорируются агрегатными функциями и не будем фильтровать данные в таблице для их исключения

select
    avg(salary_from) as avg_salary_from,
    avg(salary_to) as avg_salary_to,
    max(salary_to) as max_salary_to,
    min(salary_from) as min_salary_from
from vacancies as v

'''
# при округление числа до целого с помощью операции round в sql для типа данных float остается 0 после запятой
# поэтому операцию округления до целого произведем при выводе результатов с помощью python

In [17]:
# результат запроса
df_4_3 = pd.read_sql_query(query_4_3,connection)
print ('Среднее значение нижней границы зарплатной вилки составляет: ',round(df_4_3.iloc[0][0]), 'рублей.')
print ('Среднее значение верхней границы зарплатной вилки составляет: ',round(df_4_3.iloc[0][1]), 'рублей.')
print ('При этом минимальная нижняя граница зарплатной вилки составляет: ',round(df_4_3.iloc[0][3]), 'рублей.')
print ('Максимальная верхняя граница зарплатной вилки составляет: ',round(df_4_3.iloc[0][2]), 'рублей.')


Среднее значение нижней границы зарплатной вилки составляет:  71065 рублей.
Среднее значение верхней границы зарплатной вилки составляет:  110537 рублей.
При этом минимальная нижняя граница зарплатной вилки составляет:  30 рублей.
Максимальная верхняя граница зарплатной вилки составляет:  1000000 рублей.


4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [18]:
# текст запроса
query_4_4 = f'''
-- для получения необходимой информации воспользуемся таблице БД "vacancies"
-- для вывода в результирующей таблице, выберем данные столбцов "schedule" и "employment"
-- добавим столбец с количеством вакансий на каждое сочетание признаков 
select 
    v.schedule as рабочий_график,
    v.employment as тип_занятости,
    count(*) as количество
from vacancies as v

-- для корректной работы запроса необходимо сгруппировать данные по признакам "schedule" и "employment"
group by 
    v.schedule,
    v.employment
order by 3 desc
'''

In [19]:
# результат запроса
df_4_4 = pd.read_sql_query(query_4_4,connection)
print('Топ 10 популярных сочетаний рабочего графика и типа трудовой занятости')
df_4_4.head(10)

Топ 10 популярных сочетаний рабочего графика и типа трудовой занятости


рабочий_график        тип_занятости  количество
0       Полный день     Полная занятость       35367
1  Удаленная работа     Полная занятость        7802
2     Гибкий график     Полная занятость        1593
3  Удаленная работа  Частичная занятость        1312
4    Сменный график     Полная занятость         940
5       Полный день           Стажировка         569
6    Вахтовый метод     Полная занятость         367
7       Полный день  Частичная занятость         347
8     Гибкий график  Частичная занятость         312
9       Полный день     Проектная работа         141

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [20]:
# текст запроса
query_4_5 = f'''
-- для получения необходимой информации воспользуемся таблице БД "vacancies"
-- в качестве ответа на запрос получим данные ячеек из столбца "experience"
-- добавим к ним информацию о количестве вакансий в зависимости от опыта_работы
-- рассчитаем общее количество вакансий
with total as (
    select count(*) as cnt_vacancies
from vacancies
)

select 
    v.experience as Опыт_работы,
    count(v.id) as Число_вакансий,
    (100*count(v.id)/(select cnt_vacancies from total)) as процент -- добавим столбец с процентом от общего количества вакансий
from vacancies as v

-- для требуемых данных необходимо сгруппировать данные по признаку "experience"
group by 
    v.experience
order by 2
'''

In [21]:
# результат запроса
df_4_5 = pd.read_sql_query(query_4_5,connection)
print('Распределение количества вакансий от необходимого опыта работы с процентом от общего количества вакансий')
df_4_5

Распределение количества вакансий от необходимого опыта работы с процентом от общего количества вакансий


Опыт_работы  Число_вакансий  процент
0         Более 6 лет            1337        2
1           Нет опыта            7197       14
2       От 3 до 6 лет           14511       29
3  От 1 года до 3 лет           26152       53

***

<span style="color:#59afe1"> 

# Выводы по детальному анализу вакансий

1. Как и можно было предположить, в крупных городах представлено наибольшее количество вакансий.  
Объясняется очень просто - в крупных городах больше средств, население, а соответственно более высокие зарплаты и как   
показал дальнейший анализ, больше число зарегистрированных компаний.  
Топ 3 по количеству вакансий это:
    - Москва;
    - Санкт-Петербург;
    - Минск.

2. В базе данных 24073 вакансии имеют информацию о верхнем и нижнем уровне зарплат. Это составляет 49% от общего  
количества вакансий в БД. Я бы обратил внимание, на те 51% компаний которые не указывают ЗП. Скорее всего это компании,   
которые готовы платить за уровень, знаний, умений и компетенций соискателя. И хотя увидеть кандидата, который не гонится   
за числом в расчете, а ищет работу которая ему в первую очередь будет интересна. Но для дальнейшего анализа, это конечно минус.  
По данной БД сложно, достоверно определить средний уровень ЗП на рынке, потому что 50% информации попросту отсутствует.  
Хотя скорее всего, в 51% компаний попали, более творческие профессии. Скорее всего компания, которая ищет   
"Электромонтера контактной сети железных дорог" не задумывается по поводу того чтобы их профессия была интересна соискателю.

3. Как уже говорил выше - сложно достоверно рассуждать об уровне средних ЗП. Но из имеющихся данных можно сделать вывод о том,  
что в БД высокий разброс зарплат-возможно присутствуют аномальные значения. Для более точных значений средних зарплат, возможно,  
необходимо провести "очистку" данных и провести дополнительное исследование на:
    - характер распределения;
    - размах;
    - медиальное значение;
    - мода;
    - уточненное среднее значение.

4. Проанализировав ТОП 10 сочетаний рабочего графика и типа трудовой занятости, можно сказать о том что работодателям очень  
важна полная занятость. Соискатели на проектную занятость устраивает только 141 компанию. Что радует меня, достаточно много компаний   
готовы рассмотреть стажировку соискателя. Для дополнительного анализа можно, проверить сколько из 569 компаний предлагаю вакансии   
в сфере Data science.   
Что касается рабочего графика, то 35367 компаний хотят, чтобы будущий сотрудник очно выполнял свои обязанности.  
В то время как, удаленную работу готовы предложить только 7802 компании. Скорее всего большинство компаний не готовы к предоставлению    
удаленного формата сотрудникам, боясь, что эффективность сотрудников будет ниже чем при оффлайн формате. Дело и в том, что и многие 
сотрудники предпочитаю офисную работу, так как в офисе подходящая атмосфера для креативности и коммуникации. Хотя на мой взгляд, при  
удаленном формате можно сохранять свою эффективность и коммуникация с коллегами, рынок по прежнему скептически относится к "удаленщикам".
Но, к счастью, не рынок IT-вакансий=)

5. Распределение количества вакансий от необходимого опыта показывает, что на 53% вакансий требуются молодые специалисты с опытом от 1 до 3 лет.
Я думаю, это объяснятся тем, что на более высоких и ответственных должностях, где требования к опыту выше, у компании уже есть специалист или   
это место закрывают сотрудники из числа уже работающих на компанию, и проявивших себя. Как раз на эти освободившиеся места и требуются менее  
опытные специалисты.  
Обращая внимание, на то, что количество вакансий "без опыта" составляет 14% от всех вакансий, появляется желание сделать дополнительный анализ,   
тех компаний которые размещают эти вакансии. Возможно, это молодые компании, которые занимаются "молодыми" профессиями - SMM, блогер, копирайтер и т.д.  
Таким компаниям проще, нанять сотрудника без опыта и обучить всему с нулю.   
Скорее, всего часть этих вакансий просто не подразумевает наличия опыта: кладовщик, охранник, грузчик и т.д.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [22]:
# текст запроса
query_5_1 = f'''
-- для получения результата запроса необходимо объединить две таблицы БД: "employers" и "vacancies".
-- чтобы соотнести количество вакансий с конкретной компанией необходимо произвести объединение таблиц по ключу с id компаний

-- в качестве ответа на запрос получим имена компаний из ячеек столбца "name" таблицы "employers"
-- добавим к ним информацию о количестве вакансий этой компании, указанных в таблице "vacancies"
select
    e.name as компания, 
    count(v.id) as Число_вакансий
from vacancies as v

-- объединяем таблицы по ключу employer_id
join employers as e on v.employer_id = e.id

-- для требуемых данных необходимо сгруппировать данные по признаку "id" из таблицы "employers" 
group by    
    e.id

-- для ответа, ограничим запрос 5 первыми компаниями по количеству вакансий
order by 2 desc
limit 5
'''

In [23]:
# результат запроса
df_5_1 = pd.read_sql_query(query_5_1,connection)
print('Топ 5 компаний по количеству предоставляемых вакансий')
print(df_5_1)
print()
print('Первое место в списке по количеству вакансий -', df_5_1.iloc[0][0])
print('Пятое место в списке по количеству вакансий -', df_5_1.iloc[4][0])

Топ 5 компаний по количеству предоставляемых вакансий
        компания  Число_вакансий
0         Яндекс            1933
1     Ростелеком             491
2       Тинькофф             444
3           СБЕР             428
4  Газпром нефть             331

Первое место в списке по количеству вакансий - Яндекс
Пятое место в списке по количеству вакансий - Газпром нефть


2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [24]:
# текст запроса
query_5_2_1 = f'''
-- для получения результата запроса необходимо объединить три таблицы БД: "areas", "vacancies" и "employers".
-- так как число регионов в таблице "areas" может оказаться больше, чем указано в таблицах "employers" и "vacancies",
-- то в случае объединения методом inner join мы потеряем часть регионов которые не представлены в таблицах "employers" и "vacancies"
-- с учетом того, что в основе анализа мы рассматриваем регионы, необходимо осуществить присоединение таблиц через операцию left join без потери информации о регионах
-- к таблице "areas" присоединим таблицу "employers" по ключу area_id
-- к таблице "areas" присоединим таблицу "vacancies" по ключу area_id

-- в качестве ответа на запрос получим названия регионов из ячеек столбца "name" таблицы "area"
-- добавим к ним информацию о количестве компаний зарегистрированных в этом регионе, указанных в таблице "employers"
-- добавим к ним информацию о количестве вакансий в этом регионе, указанных в таблице "vacancies"
-- учтем, что после соединения 3 таблиц количество ячеек размножится - 
-- например, в Батайске зарегистрировано 3 компании и размещено 8 вакансий. Результатом объединения будет 24 строки
-- для того, чтобы избежать дубликатов при расчете количества воспользуемся операцией DISTINCT  
select 
    a.name as Регион,
    count (distinct e.id) as число_компаний,
    count (distinct v.id) as число_вакансий
from areas as a

-- присоединяем таблицу "vacancies"
left join vacancies as v on v.area_id = a.id

-- присоединяем таблицу "employers"
left join employers as e on e.area = a.id

-- для требуемых данных необходимо сгруппировать данные по признаку "id" из таблицы "areas" 
group by 
    a.id

-- для анализа, ограничим запрос 20 первыми компаниями по количеству компаний
order by 2 desc
limit 10
'''

In [25]:
df_5_2_1 = pd.read_sql_query(query_5_2_1,connection)
print('Топ 10 регионов по количеству зарегистрированных компаний и число вакансий в этих регионах')
print(df_5_2_1)

Топ 10 регионов по количеству зарегистрированных компаний и число вакансий в этих регионах
            Регион  число_компаний  число_вакансий
0           Москва            5864            5333
1  Санкт-Петербург            2217            2851
2            Минск            1115            2112
3           Алматы             721            1892
4     Екатеринбург             609            1698
5      Новосибирск             573            2006
6           Казань             480            1415
7  Нижний Новгород             426            1670
8           Россия             410               0
9        Краснодар             409            1301


In [26]:
# составим запрос для поиска региона в котором наибольшее количество зарегистрированных компаний при отсутствие вакансий
# для составления запроса будет придерживаться той же логики что и для запроса query_5_2_1
query_5_2_2 = f'''
-- в качестве ответа на запрос получим названия региона из ячеек столбца "name" таблицы "area"
-- добавим к нему информацию о количестве компаний зарегистрированных в этом регионе, указанных в таблице "employers"
select 
    a.name as Регион,
    count (distinct e.id) as число_компаний
from areas as a

-- присоединяем таблицу "vacancies"
left join vacancies as v on v.area_id = a.id

-- присоединяем таблицу "employers"
left join employers as e on e.area = a.id

-- выбираем только те регионы в которых отсутствуют вакансии
where v.id is null

-- для требуемых данных необходимо сгруппировать данные по признаку "id" из таблицы "areas" 
group by 
    a.id

-- для ответа на задние необходим ТОП 1 среди регионов по количеству зарегистрированных компаний
-- поэтому ограничим запрос 1 строкой
order by 2 desc
limit 1
'''

In [27]:
# результат запроса
df_5_2_2 =pd.read_sql_query(query_5_2_2,connection)
print('На первом месте по числу зарегистрированных компаний среди регионов, о которых нет информации в таблице "vacancies" находится', df_5_2_2.iloc[0][0]+'.')
print('В этом регионе зарегистрировано',df_5_2_2.iloc[0][1], 'вакансий.')

На первом месте по числу зарегистрированных компаний среди регионов, о которых нет информации в таблице "vacancies" находится Россия.
В этом регионе зарегистрированно 410 вакансий.


3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [28]:
# текст запроса
query_5_3 = f'''
-- для получения результата запроса необходимо объединить две таблицы БД: "vacancies" и "employers".
-- так как число компаний в таблице "employers" может оказаться больше, чем содержится в таблице "vacancies",
-- то в случае объединения методом inner join мы потеряем часть компаний которые не представлены в таблице "vacancies" (хотя если рассматривать ТОП, то это не принципиально)
-- с учетом того, что в основе анализа мы рассматриваем компании, необходимо осуществить присоединение через операцию left join
-- к таблице "employers" присоединим таблицу "vacancies" по ключу employer_id

-- в качестве ответа на запрос получим названия компаний из ячеек столбца "name" таблицы "employers"
-- добавим к ним информацию о количестве регионов в которых компания публикует свои вакансии, указанных в таблице "vacancies"
select
    e.name as Компания,
    Count(distinct v.area_id) as Число_регионов
from employers as e
left join vacancies as v on v.employer_id = e.id

-- для требуемых данных необходимо сгруппировать данные по признаку "name" из таблицы "employers" 
group by 
    e.name
order by 2 desc
'''

In [29]:
# результат запроса
df_5_3 = pd.read_sql_query(query_5_3,connection)
print('Топ 5 компаний по числу регионов в которых компания публикует свои вакансии')
df_5_3.head(5)

Топ 5 компаний по числу регионов в которых компания публикует свои вакансии


Компания  Число_регионов
0                  Яндекс             181
1              Ростелеком             152
2              Спецремонт             116
3  Поляков Денис Иванович              88
4               ООО ЕФИН               71

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [30]:
# текст запроса
query_5_4 = f'''
-- для получения результата запроса необходимо объединить две таблицы БД: "employers_industries" и "employers".
-- так как число компаний в таблице "employers" может оказаться больше, чем содержится в таблице "employers_industries",
-- то в случае объединения методом inner join мы потеряем часть компаний которые не представлены в таблице "employers_industries"
-- с учетом того, что в основе анализа мы рассматриваем компании, необходимо осуществить присоединение через операцию left join
-- к таблице "employers" присоединим таблицу "employers_industries" по ключу employer_id

-- выберем из таблицы количество компаний, у которых не указана сфера деятельности
-- для того чтобы ответ на SQL-запрос был более информативным, добавим общее количество компаний и выведем результат в виде таблицы

-- в первой строке таблицы получим информацию о количестве компаний, у которых не указана сфера деятельности
select 
    'Без сферы деятельности' as type,
    count(*) as cnt_empl_without_industries
from employers as e
left join employers_industries as e_i on e.id = e_i.employer_id

-- пропишем условие для отбора компаний, у которых не указана сфера деятельности
where industry_id is null

-- объединяем две строки с помощью операции UNION
union

-- во второй строке получим общее количество компаний
select 
    'Общее число',
    count(*) as total_cnt
from employers as e

'''

In [31]:
# результат запроса
df_5_4 = pd.read_sql_query(query_5_4,connection)
print(df_5_4.iloc[0][1],'компаний не указали сферу своей деятельности.')
print('Это составляет',round(100*df_5_4.iloc[0][1]/df_5_4.iloc[1][1]),'% от общего количества компаний.')

8419 компаний не указали сферу своей деятельности.
Это составляет 36 % от общего количества компаний.


5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [32]:
# текст запроса
query_5_5 = f'''
-- для получения результата запроса необходимо объединить две таблицы БД: "employers_industries" и "employers".
-- так как число компаний в таблице "employers" может оказаться больше, чем содержится в таблице "employers_industries",
-- то в случае объединения методом inner join мы потеряем часть компаний которые не представлены в таблице "employers_industries"
-- с учетом того, что в основе анализа мы рассматриваем компании, необходимо осуществить присоединение через операцию left join
-- к таблице "employers" присоединим таблицу "employers_industries" по ключу employer_id

-- выберем из таблицы имена компаний, у которых указано 4 сферы деятельности
-- для того чтобы ответ на SQL-запрос был более информативным, добавим общее количество компаний и выведем результат в виде таблицы

select 
    e.name
from employers as e
left join employers_industries as e_i on e.id = e_i.employer_id

-- для требуемых данных необходимо сгруппировать данные по признаку "id" из таблицы "employers" 
group by   
    e.id

-- пропишем условие для отбора компаний, у которых указано 4 сферы деятельности
having count(distinct e_i.industry_id) = 4

-- осуществим сортировку в алфавитном порядке
order by 1
'''

In [33]:
# результат запроса
df_5_5 = pd.read_sql_query(query_5_5,connection)
print(df_5_5.iloc[2,0], 'находится на третьем месте в алфавитном списке компаний, у которых указано четыре сферы деятельности')

2ГИС находится на третьем месте в алфавитном списке компаний, у которых указано четыре сферы деятельности


6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [34]:
# текст запроса
query_5_6 = f'''
-- для получения результата запроса необходимо объединить две таблицы БД: "employers_industries" и "industries".
-- к таблице "employers_industries" присоединим таблицу "industries" по ключу industry_id
-- для того чтобы ответ на SQL-запрос был более информативным, добавим общее количество компаний и выведем результат в виде таблицы

-- в первой строке таблицу укажем число компаний, которые в качестве сферы деятельности указали 'Разработка программного обеспечения'
select 
    'Develop-компании' as type, 
    count(*)
from employers_industries as e_i
join industries as i on e_i.industry_id = i.id
where i.name = 'Разработка программного обеспечения'

-- объединяем две строки с помощью операции UNION
union

-- во второй строке получим общее количество компаний
--для этого воспользуемся таблицей 'employers'
select 
    'Общее число компаний',
    count(*)
from employers as e

'''

In [35]:
# результат запроса
df_5_6 = pd.read_sql_query(query_5_6,connection)
print(df_5_6.iloc[0][1],'компаний в качестве сферы деятельности указывает "Разработка программного обеспечения".')
print('Всего',round(100*df_5_6.iloc[0][1]/df_5_6.iloc[1][1]),'% процентов компаний занимаются разработкой программного обеспечения.')

3553 компаний в качестве сферы деятельности указывает "Разработка программного обеспечения".
Всего 15 % процентов компаний занимаются разработкой программного обеспечения.


7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [36]:
# код для получения списка городов-милионников
import requests # импорт библиотеки для работы с веб-запросами
from bs4 import BeautifulSoup #импорт библиотеки для работы с html страницами
url = 'https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8'
response = requests.get(url) # Отправляем запрос на получение данных с веб-страницы
page = BeautifulSoup(response.text,'html.parser') # Преобразуем полученные данные в html-текст
table_cites = page.find('table', class_="standard sortable") # При помощи необходимого хештега ищем таблицу с городами-миллиониками

# Преобразуем полученную информацию в DataFrame
headers =[] # Создаем список будущих заголовков таблицы
# Заполняем список заголовками таблицы с помощью цикла 
for th in table_cites.find_all('th'):
    headers.append(th.text)
# По полученным заголовка формируем заготовку под DataFrame
my_table_cites = pd.DataFrame(columns = headers)
# Заполняем DataFrame данными
for tr in table_cites.find_all('tr')[1:]: # Внутри хештега <tr> - находится данные строки таблицы
    row = [td.text for td in tr.find_all('td')] # Внутри хештега <td> - находится данные ячеек таблицы. Заполняем строку данными из ячеек. 
    length = len(my_table_cites)
    my_table_cites.loc[length] = row
list_cites = my_table_cites['Город'].to_list() #получим список городов-милионников

In [37]:
# Создадим строку (шаблон) со списком городов-милионников  и условием для фильтрации sql-запроса [areas.name in ("Название города 1","Название города 2",...)]
str_cites = 'a.name in ('
for i in list_cites:
    str_cites+="'"+i+"'"+","
str_cites = str_cites[:-1]+')' # добавляем скобку в конец и убираем лишнюю запятую

# Формируем текст запроса

# Начало запроса
# для получения результата запроса необходимо объединить три таблицы БД: "employers", "vacancies" и "areas".
# так как нам нужно получить подробную информацию только о компании "Яндекс", то можно использовать метод объединения inner join
# к таблице "employers" присоединим таблицу "vacancies" по ключу employer_id
# к таблице "vacancies" присоединим таблицу "areas" по ключу area_id
# по заданию, добавим строку Total с общим количеством вакансий компании в этих регионах
query_5_7_0 =f'''
select
    e.name as Компания,
    a.name as Город,
    count(a.name) as Число_вакансий
from employers as e
join vacancies as v on e.id = v.employer_id
join areas as a on v.area_id = a.id
where e.name = 'Яндекс' and
'''
# продолжение 1
query_5_7_1 = f'''
group by 
    e.name,
    a.name
union
select
    e.name,
    'total' as total,
    count(a.id)
from employers as e
join vacancies as v on e.id = v.employer_id
join areas as a on v.area_id = a.id
where e.name = 'Яндекс' and
'''
# продолжение 2
query_5_7_2 = f'''
group by 
    e.name
order by 3 desc
'''

# текст запроса
query_5_7 = query_5_7_0 + str_cites + query_5_7_1 + str_cites + query_5_7_2

In [38]:
# результат запроса
df_5_7 = pd.read_sql_query(query_5_7,connection)
print('Топ городов-миллионников по количеству вакансий от компании "Яндекс"')
df_5_7

Топ городов-миллионников по количесвту вакансий от компании "Яндекс"


Компания            Город  Число_вакансий
0    Яндекс            total             485
1    Яндекс           Москва              54
2    Яндекс  Санкт-Петербург              42
3    Яндекс     Екатеринбург              39
4    Яндекс  Нижний Новгород              36
5    Яндекс      Новосибирск              35
6    Яндекс          Воронеж              32
7    Яндекс        Краснодар              30
8    Яндекс           Самара              26
9    Яндекс              Уфа              26
10   Яндекс            Пермь              25
11   Яндекс           Казань              25
12   Яндекс   Ростов-на-Дону              25
13   Яндекс        Волгоград              24
14   Яндекс       Красноярск              23
15   Яндекс        Челябинск              22
16   Яндекс             Омск              21

***

<span style="color:#59afe1"> 

# Выводы по анализу работодателей

Как мы можем увидеть в анализе, в ТОП 5 компаний по числу вакансий входят:  

    1) Яндекс - 1933 вакансии;
    2) Ростелеком - 491 вакансия;
    3) Тинькофф - 444 вакансии;
    4) СБЕР - 428 вакансий;
    5) Газпром нефть - 331 вакансия.

Я думаю это связано с тем, что это крупные компании с широкой географией вакансий.  
Яндекс и Ростелеком является крупнейшими производителями IT-продуктов в России, поэтому неудивительно,    
что ему требуется большое число сотрудников и у него есть возможность их нанимать благодаря  
предоставления удаленного формата работы. В подтверждение этого, Яндекс и Ростелеком находятся на первом  
и втором месте соответственно, по числу регионов в которых эти компании предоставляют свои вакансии.  
Яндекс разместил свои вакансии для 181 региона, а Ростелеко в 152 регионах. Тинькофф, находясь на 14 месте  
в этом списке, разместил свои вакансии в 43 регионах и имеет те же причины попадания в ТОП. Не смотря на то,  
что СБЕР и Газпром нефть находятся на 35 и 36 местах в рейтинге разместив свои вакансии в 24 регионах,  
корреляция между рейтингом компаний по числу размещенных вакансий и рейтингом компаний по числу регионов   
в которых эти компании разместили свои вакансии очевидна.  
Помимо всего прочего, СБЕР и Газпром нефть одни из самых крупных компаний в России  
с физическим представительством в 24 регионах, очевидно что им нужно большое количество сотрудников.  
Все это объясняет их попадание в ТОП.


# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [39]:
# текст запроса
query_6_1 =f'''
-- для получения необходимой информации воспользуемся таблице БД "vacancies"
-- в качестве ответа на запрос посчитаем количество строк таблицы при соблюдение условий: 
-- значение ячейки в столбце c именем вакансии содержит "data" или "данн"
-- для реализации воспользуемся оператором like и значком % (искомые слова могу находится в тексте на любой позиции)
select
    'DS-вакансии' as type,
    count(*)
from vacancies as v
where lower(name) like '%data%' or lower(name) like '%данн%'

-- для сравнения добавим строку с общем количеством вакансий в базе данных
union
select 
    'Общее число вакансий',
    count(*)
from vacancies as v

'''

In [40]:
# результат запроса
df_6_1 = pd.read_sql_query(query_6_1,connection)
print('Число вакансий имеющий отношений к анализу и обработке данных :',df_6_1.iloc[0][1])
print('Это составляет',round(100*df_6_1.iloc[0][1]/df_6_1.iloc[1][1]),'% от общего количества вакансий в БД.')



Число вакансий имеющий отношений к анализу и оброботке данных : 1771
Это составляет 4 % от общего количества вакансий в БД.


2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [41]:
# текст запроса
query_6_2 = f'''
-- для получения необходимой информации воспользуемся таблице БД "vacancies"
-- в качестве ответа на запрос посчитаем количество строк таблицы при соблюдение условий задания 
-- для реализации воспользуемся оператором like и значком %(искомые слова могу находится в тексте на любой позиции)
-- обратим внимание на то, что для поиска вакансий с названием включающем 'ML'(Machine Learning) необходимо сохранить регистр букв без изменений
-- в противном случае мы можем получить ложные значение, такие например как 'teaMLead'
-- так же обратим внимание, что необходимо "очистить" данные от таких вакансий которые содержат 'C++ разработчик(QML)'
select 
    'DS-junior' as type,
    count(*) as Число_вакансий
from vacancies as v
where 
    (lower(name) like '%data scientist%' 
    or lower(name) like '%data science%' 
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and lower(name) not like '%html%' and lower(name) not like '%qml%')
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%%'
    )
and
    (lower(name) like '%junior%' 
    or experience = 'Нет опыта'
    or employment = 'Стажировка'
    )
union
select 
    'Общее число вакансий',
    count(*)
from vacancies as v
'''

In [42]:
# результат запроса
df_6_2 = pd.read_sql_query(query_6_2,connection)
print('Для начинающего Data science специалиста в БД представлено',df_6_2.iloc[0][1],'вакансия.')
print('Это составляет',round(100*df_6_2.iloc[0][1]/df_6_2.iloc[1][1],2),'% от общего числа вакансий.')

Для начинающего Data science специалиста в БД представлено 51 вакансия.
Это составляет 0.1 % от общего числа вакансий.


3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [43]:
# текст запроса
# за основу текущего запроса примем запрос query_6_2
query_6_3 = f'''
select 
    'DS-SQL' as type,
    count(*)
from vacancies as v
where 
    (
    lower(name) like '%data scientist%' 
    or lower(name) like '%data science%' 
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and lower(name) not like '%html%' )
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'
    )
and 
    (
    lower(key_skills) like '%sql%'
    or lower(key_skills) like '%postgres%'
    )
union
select 
    'Общее число DS',
    count(*)
from vacancies as v
where 
    (
    lower(name) like '%data scientist%' 
    or lower(name) like '%data science%' 
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and lower(name) not like '%html%' )
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'
    )
'''

In [44]:
# результат запроса
df_6_3 = pd.read_sql_query(query_6_3,connection)
print('У',df_6_3.iloc[0][1],' вакансии в качестве ключевых навыков указан "SQL" или "postgres".')
print('Это составляет',round(100*df_6_3.iloc[0][1]/df_6_3.iloc[1][1]) ,'% от общего числа вакансий для Data Science разработчиков.')

У 201  вакансии в качестве ключевых навыков указан "SQL" или "postgres".
Это составляет 42 % от общего числа вакансий для Data Science разработчиков.


4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [45]:
# текст запроса
# за основу текущего запроса примем запрос query_6_2
query_6_4 = f'''
select 
    'DS-Python' as type,
    count(*)
from vacancies as v
where 
    (
    lower(name) like '%data scientist%' 
    or lower(name) like '%data science%' 
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and lower(name) not like '%html%' )
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'
    )
and 
    lower(key_skills) like '%python%'
union
select 
    'Общее число DS',
    count(*)
from vacancies as v
where 
    (
    lower(name) like '%data scientist%' 
    or lower(name) like '%data science%' 
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and lower(name) not like '%html%' )
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'
    )
'''

In [46]:
# результат запроса
df_6_4 = pd.read_sql_query(query_6_4,connection)
print('Число вакансий для Data Science специалистов с требованием к знанию Python:',df_6_4.iloc[0][1])
print('Это',round(100*df_6_4.iloc[0][1]/df_6_4.iloc[1][1]),'% от общего числа вакансий для Data Science разработчиков.')

Число вакансий для Data Science специалистов с требованием к знанию Python: 351
Это 73 % от общего числа вакансий для Data Scince разработчиков.


5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [47]:
# текст запроса
query_6_5 = f'''
-- для получения необходимой информации воспользуемся таблице БД "vacancies"
-- обратим вниманием, что ключевые навыки разделены знаком табуляции
-- поэтому, для подсчета числа ключевых навыков в конкретной вакансии нужно посчитать число знаков табуляции
-- число ключевых навык на единицу больше числа знаков табуляции
select 
    round(avg(length(key_skills) - length(replace(key_skills,chr(9),'')) + 1),2)
from vacancies as v
where 
    (
    lower(name) like '%data scientist%' 
    or lower(name) like '%data science%' 
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and lower(name) not like '%html%' )
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'
    )
and key_skills is not null
'''

In [48]:
# результат запроса
df_6_5 = pd.read_sql_query(query_6_5,connection)
print('В среднем для вакансии Data Science специалиста указывается ',round(df_6_5.iloc[0][0]),'ключевых навыков')
# округление произвел до целого числа, потому что показалось странным дробное число вакансий

В среднем для вакансии Data Scince специалиста указывается  6 ключевых навыков


6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [49]:
# текст запроса
query_6_6 = f'''
-- для получения необходимой информации воспользуемся таблице БД "vacancies"
-- обратим внимание на то что в задание указано нужно найти среднюю зарплату для двух случаев
-- для реализации этого условия задания воспользуемся оператором case..when..then..else..end 
select
    experience as Опыт_работы,
    round(
    avg(
    case
        when salary_from is not null and salary_to is not null -- если заполнены оба полей зарплатной вилки
        then (salary_to+salary_from)/2 -- рассчитываем ЗП в случае выполнения условия
        else (coalesce(salary_to,0)+coalesce(salary_from,0)) -- рассчитываем ЗП в случае если условие не выполняется (учитываем, что есть условие на всю таблицу)
    end
    )) as Средняя_зарплата  
from vacancies as v
where 
    (
    lower(name) like '%data scientist%' 
    or lower(name) like '%data science%' 
    or lower(name) like '%исследователь данных%'
    or (name like '%ML%' and lower(name) not like '%html%' )
    or lower(name) like '%machine learning%'
    or lower(name) like '%машинн%обучен%'
    )
and 
    (salary_from is not null or salary_to is not null)
group by 
    experience
order by 2 desc
'''

In [50]:
# результат запроса
df_6_6 = pd.read_sql_query(query_6_6,connection)
print('Распределение средней заработной платы Data Science специалиста в зависимости от опыта работы')
print(df_6_6)
print()
print('C опытом работ от 3 до 6 лет Data Science специалист может рассчитывать на среднюю заработную плату равную',round(df_6_6[df_6_6['Опыт_работы']=='От 3 до 6 лет'].iloc[0][1]),'рублей.')

Распределение средней заработной платы Data Science специалиста в зависимотси от опыта работы
          Опыт_работы  Средняя_зарплата
0       От 3 до 6 лет          243115.0
1  От 1 года до 3 лет          139675.0
2           Нет опыта           74643.0

C опытом работ от 3 до 6 лет Data Science специалист может расчитывать на среднюю заработную плату равную 243115 рублей.


***

<span style="color:#59afe1"> 

# Выводы по предметному анализу

В базе данных 1771 вакансия имеют отношение к Data Science специалистам. Это 4% от общего количества вакансий.    
Из них 51 вакансия для специалистов уровня Junior. Это 3% от вакансий для Data Science специалистов.  
С учетом того, что в БД 53% вакансий требуют опыт работы от 1 года до 3 и всего 14% без опыта работы -   
можно сделать вывод о том, что начинающим специалистам в Data Science будет не лего сделать свой первый шаг.  
По крайней мере, использую базу данных вакансий Head Hunter.   
По этому, чтобы увеличить свои шансы на успешный поиск работы в области Data Science, будущему специалисту  
необходимо в совершенстве освоить следующие hard skills:  
    - Python. 73% компаний, указывают навыке в своем списке ключевых навыков;  
    - "SQL" и "postgres". По статистике 42% вакансий Data Science специалистов, требуют эти навыки в резюме соискателя;  
    - По мимо этого, необходимо обогатить свое резюме еще не менее 4 hard skills - например, Machine Learning, Deep Learning, Computer Vision, Pandas, Numpy, Big Data и т.д.  
Все это не напрасно, в случае успешного собеседования. Начинающему специалисту на старте его работы обещают    
в среднем 75000 рублей, Специалисту с опытом работы от года до трех 140000 рублей, а более опытным Data Science  
разработчикам предлагаю зарплаты в среднем 240000 рублей.

Сам себе очень рекомендую уйти в профессию Data Science разработчика =)



# Общий вывод по проекту

В России, и странах СНГ прослеживается тренд на IT специальности. Такие компании как Яндекс, Ростелеком и СБЕР бьют все ТОПы по количеству предоставляемых вакансий. Хотя если брать направления Data Science разработчиков, то понятно - сегодня нельзя сказать, что рынок в первую очередь нуждается именно в этих специалистах. Но с учетом, развитие отрасли, предлагаемого уровня зарплат и ростом популярности IT направлений можно сказать, что не за горами находится восходящий тренд для всей IT индустрии в России.  